src/backend/PipelineLayout.cpp

```diff
@@ -20,10 +20,30 @@
 
 namespace backend {
 
+    MaybeError ValidatePipelineLayoutDescriptor(DeviceBase*, const nxt::PipelineLayoutDescriptor* descriptor) {
+        NXT_TRY_ASSERT(descriptor->nextInChain == nullptr, "nextInChain must be nullptr");
+        NXT_TRY_ASSERT(descriptor->numBindGroupLayouts <= kMaxBindGroups, "too many bind group layouts");
+        for (uint32_t i = 0; i < descriptor->numBindGroupLayouts; ++i) {
+            NXT_TRY_ASSERT(descriptor->bindGroupLayouts[i].Get() != nullptr, "bind group layouts may not be null");
+        }
+        return {};
+    }
+
     // PipelineLayoutBase
 
-    PipelineLayoutBase::PipelineLayoutBase(PipelineLayoutBuilder* builder)
-        : mBindGroupLayouts(std::move(builder->mBindGroupLayouts)), mMask(builder->mMask) {
+    PipelineLayoutBase::PipelineLayoutBase(DeviceBase* device, const nxt::PipelineLayoutDescriptor* descriptor) {
+        ASSERT(descriptor->numBindGroupLayouts <= kMaxBindGroups);
+        for (uint32_t group = 0; group < descriptor->numBindGroupLayouts; ++group) {
+            mBindGroupLayouts[group] = reinterpret_cast<BindGroupLayoutBase*>(descriptor->bindGroupLayouts[group].Get());
+            mMask.set(group);
+        }
+        for (uint32_t group = descriptor->numBindGroupLayouts; group < kMaxBindGroups; ++group) {
+            auto builder = device->CreateBindGroupLayoutBuilder();
```

Comment:
Is this needed? Parts of the CL, for example `CommandBufferStateTracker.cpp` changes seem to assume no.

Reply:
There is still some work left to be able to remove this. I'll leave a TODO to fix it.